src/RequestPool.cpp

```diff
@@ -13,54 +13,75 @@ auto RequestPool::Reserve(
                 *this,
                 "",
                 std::chrono::milliseconds { 0 },
+                std::nullopt,
                 [](RequestHandle r) { (void)r; }));
         m_requests.emplace_back(std::move(request_handle_ptr));
     }
 }
 
 auto RequestPool::Produce(
-    const std::string& url) -> RequestHandle
+    std::string_view url) -> RequestHandle
 {
     using namespace std::chrono_literals;
     return Produce(url, nullptr, 0ms);
 }
 
 auto RequestPool::Produce(
-    const std::string& url,
-    std::chrono::milliseconds timeout) -> RequestHandle
+    std::string_view url,
+    std::chrono::milliseconds connection_timeout) -> RequestHandle
 {
-    return Produce(url, nullptr, timeout);
+    return Produce(url, nullptr, connection_timeout);
 }
 
 auto RequestPool::Produce(
-    const std::string& url,
+    std::string_view url,
     std::function<void(RequestHandle)> on_complete_handler,
-    std::chrono::milliseconds timeout) -> RequestHandle
+    std::chrono::milliseconds connection_timeout) -> RequestHandle
+{
+    return Produce(url, std::move(on_complete_handler), connection_timeout, std::nullopt);
+}
+
+auto RequestPool::Produce(
+    std::string_view url,
+    std::chrono::milliseconds connection_timeout,
+    std::chrono::milliseconds response_wait_time) -> RequestHandle
+{
+    return Produce(url, nullptr, connection_timeout, response_wait_time);
+}
+
+
+auto RequestPool::Produce(
+    std::string_view url,
+    std::function<void(RequestHandle)> on_complete_handler,
+    std::chrono::milliseconds connection_timeout,
+    std::optional<std::chrono::milliseconds> response_wait_time) -> RequestHandle
 {
-    m_lock.lock();
 
+    m_lock.lock();
```

Comment:
this wasn't your code, but we should still use some RAII locking mechanism for safety. The unique_lock for example has an unlock method that let's you release the lock early if you desire, while not introducing the possibility for code paths that never unlock